server/app/services/apikey/ApiKeyService.java

```diff
@@ -0,0 +1,278 @@
+      throws ProgramNotFoundException {
+    ImmutableSet<String> programSlugs = programService.getAllProgramSlugs();
+
+    for (String formDataKey : form.rawData().keySet()) {
+      Matcher matcher = GRANT_PROGRAM_READ_PATTERN.matcher(formDataKey);
+
+      if (!matcher.find()) {
+        continue;
+      }
+
+      matcher.matches();
+
+      String programSlug = matcher.group(1);
+
+      if (!programSlugs.contains(programSlug)) {
+        throw new ProgramNotFoundException(programSlug);
+      }
+
+      apiKey.getGrants().grantProgramPermission(programSlug, Permission.READ);
+    }
+
+    return form;
+  }
+
+  private String generateSecret(int length) {
+    KeyGenerator keyGen;
+
+    try {
+      keyGen = KeyGenerator.getInstance("AES");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+
+    keyGen.init(length);
+
+    byte[] secret = keyGen.generateKey().getEncoded();
+    return Base64.getEncoder().encodeToString(secret);
+  }
+
+  private String salt(String message) {
+    byte[] rawMessage = Base64.getDecoder().decode(message);
+    byte[] rawKey = Base64.getDecoder().decode(secretSalt);
+
+    HashFunction hashFunction = Hashing.hmacSha256(rawKey);
+    HashCode saltedMessage = hashFunction.hashBytes(rawMessage);
+
+    return Base64.getEncoder().encodeToString(saltedMessage.asBytes());
+  }
+
+  private String getAuthorityId(CiviFormProfile profile) {
+    String authorityId = profile.getAuthorityId().join();
+
+    if (authorityId != null) {
+      return authorityId;
+    }
+
+    if (environment.isDev()) {
+      return "dev-admin";
+    }
+
+    throw new RuntimeException("ApiKey creator must have authority_id.");
+  }
+
+  /** Holds state relevant to the result of attempting to create an {@link ApiKey}. */
+  public static class ApiKeyCreationResult {
+    private final Optional<ApiKey> apiKey;
+    private final Optional<DynamicForm> form;
+    private final Optional<String> credentials;
+
+    /** Constructs an instance in the case of success. */
+    public static ApiKeyCreationResult success(ApiKey apiKey, String credentials) {
+      return new ApiKeyCreationResult(
+          Optional.of(apiKey), Optional.of(credentials), Optional.empty());
+    }
+
+    /** Constructs an instance in the case of failure. */
+    public static ApiKeyCreationResult failure(DynamicForm form) {
+      return new ApiKeyCreationResult(Optional.empty(), Optional.empty(), Optional.of(form));
+    }
+
+    private ApiKeyCreationResult(
+        Optional<ApiKey> apiKey, Optional<String> credentials, Optional<DynamicForm> form) {
+      this.apiKey = apiKey;
+      this.credentials = credentials;
+      this.form = form;
+    }
+
+    /** Returns true for the key was created. */
+    public boolean successful() {
+      return credentials.isPresent();
+    }
+
+    /** Returns the API key if successful, throws {@code NullPointerException} if not. */
```

Comment:
by declared I meant a non RuntimeException

Reply:
Hmm how about instead of letting the optional nullpointerexeption throw we instead throw a runtimexception with a message like "cannot retrieve credentials/apiKey from a failed ApiKeyMutationResult"?

Reply:
added to the class-level javadoc